rust/src/http2/huffman.rs

```diff
@@ -482,23 +484,44 @@ fn http2_huffman_table_len30(n: u32) -> Option<u8> {
     }
 }
 
-named!(http2_decode_huffman_len30<(&[u8], usize), u8>,
-    complete!( map_opt!(take_bits!(30u32), http2_huffman_table_len30))
-);
+fn http2_decode_huffman_len30(input: (&[u8], usize)) -> IResult<(&[u8], usize), u8> {
+    complete(map_opt(take_bits(30u32), http2_huffman_table_len30))(input)
+}
 
 //hack to end many0 even if some bits are remaining
 fn http2_decode_huffman_end(input: (&[u8], usize)) -> IResult<(&[u8], usize), u8> {
-    return Err(Err::Error((input, ErrorKind::Eof)));
+    return Err(Err::Error(make_error(input, ErrorKind::Eof)));
 }
 
 //we could profile and optimize performance here
-named!(pub http2_decode_huffman<(&[u8], usize), u8>,
-    alt!(http2_decode_huffman_len5 | http2_decode_huffman_len6 | http2_decode_huffman_len7 |
-    http2_decode_huffman_len8 | http2_decode_huffman_len10 | http2_decode_huffman_len11 |
-    http2_decode_huffman_len12 | http2_decode_huffman_len13 | http2_decode_huffman_len14 |
-    http2_decode_huffman_len15 | http2_decode_huffman_len19 | http2_decode_huffman_len20 |
-    http2_decode_huffman_len21 | http2_decode_huffman_len22 | http2_decode_huffman_len23 |
-    http2_decode_huffman_len24 | http2_decode_huffman_len25 | http2_decode_huffman_len26 |
-    http2_decode_huffman_len27 | http2_decode_huffman_len28 | http2_decode_huffman_len30 |
-    http2_decode_huffman_end)
-);
+pub fn http2_decode_huffman(input: (&[u8], usize)) -> IResult<(&[u8], usize), u8> {
+    // trait nom::branch::Alt is implemented for lists up to size 20,
+    // so use nested `alt` as a workaround (see nom documentation for `alt`)
+    alt((
+        http2_decode_huffman_len5,
+        http2_decode_huffman_len6,
+        http2_decode_huffman_len7,
+        http2_decode_huffman_len8,
+        http2_decode_huffman_len10,
+        http2_decode_huffman_len11,
+        http2_decode_huffman_len12,
+        http2_decode_huffman_len13,
+        http2_decode_huffman_len14,
+        http2_decode_huffman_len15,
+        http2_decode_huffman_len19,
+        http2_decode_huffman_len20,
+        http2_decode_huffman_len21,
+        http2_decode_huffman_len21,
+        http2_decode_huffman_len22,
+        http2_decode_huffman_len23,
+        http2_decode_huffman_len24,
+        http2_decode_huffman_len25,
+        http2_decode_huffman_len26,
+        http2_decode_huffman_len27,
+        alt((
```

Comment:
Why another alt ?

Reply:
This is detailed in the comment above the big `alt` statement: nom has implemented the `Alt` trait (used internally by the `alt` function) only for lists up to size 20, and this list has 21 elements :)
Using nested `alt` is the proposed workaround from the nom documentation (https://docs.rs/nom/7.1.0/nom/branch/fn.alt.html):
```
It takes as argument a tuple of parsers. There is a maximum of 21 parsers. If you need more, it is possible to nest them in other alt calls, like this: alt(parser_a, alt(parser_b, parser_c))

```

https://github.com/chifflier/suricata/blob/dev/nom7-upgrade-http2-v0.1/rust/src/http2/huffman.rs#L497-L499

Reply:
Thanks Pierre for the comment I had not seen 🙄